File: 207-course-schedule/course-schedule.cpp
class Solution {
public:
    bool canFinish(int n, vector<vector<int>>& preq) {
        
        vector<int> adj[n];
        vector<int> degree(n,0);

        for(int i=0;i<preq.size();i++){
            adj[preq[i][0]].push_back(preq[i][1]);
            degree[preq[i][1]]++;
        }
        vector<int> bfs;
        
        for(int i=0;i<n;i++){
            if(degree[i]==0){bfs.push_back(i);}
        }

        for(int i=0;i<bfs.size();i++){
            int ind = bfs[i];
            for(int j=0;j<adj[ind].size();j++){
                int num = adj[ind][j];
                degree[num]--;
                if(degree[num]==0){bfs.push_back(num);}
            }
        }

        return bfs.size()==n;
    }
};

File: 207-course-schedule/course-schedule.java
class Solution {

    public boolean canFinish(int numc, int[][] prereq) {
        List<List<Integer>> li = new ArrayList<>();
        int degree[] = new int[numc];
        List<Integer> bfs = new ArrayList<>();

        for(int i=0;i<numc;i++){
            li.add(new ArrayList<>());
        }

        for(int i=0;i<prereq.length;i++){
            degree[prereq[i][0]]++;
            li.get(prereq[i][1]).add(prereq[i][0]);
        }

        for(int i=0;i<numc;i++){
            if(degree[i]==0){bfs.add(i);}
        }

        for(int i=0;i<bfs.size();i++){
            int ind = bfs.get(i);
            for(int num: li.get(ind)){
                degree[num]--;
                if(degree[num]==0){bfs.add(num);}
            }
        }
        return bfs.size()==numc;
    }
}